src/data/finset/gcd.lean

```diff
@@ -0,0 +1,129 @@
+/-
+Copyright (c) 2020 Aaron Anderson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Aaron Anderson
+-/
+import data.finset.fold
+import data.multiset.gcd
+
+/-!
+# GCD and LCM operations on finsets
+
+TODO: simplify with a tactic and `data.finset.lattice`
+-/
+
+variables {α β γ : Type*} [comm_cancel_monoid_with_zero α] [nontrivial α] [gcd_monoid α]
+
+namespace finset
+open multiset
+
+/-! ### lcm -/
+section lcm
+
+/-- Least common multiple of a finite set -/
+def lcm (s : finset β) (f : β → α) : α := s.fold gcd_monoid.lcm 1 f
+
+variables {s s₁ s₂ : finset β} {f : β → α}
+
+lemma lcm_def : s.lcm f = (s.1.map f).lcm := rfl
+
+@[simp] lemma lcm_empty : (∅ : finset β).lcm f = 1 :=
+fold_empty
+
+@[simp] lemma lcm_dvd_iff {a : α} : s.lcm f ∣ a ↔ (∀b ∈ s, f b ∣ a) :=
+begin
+  apply iff.trans multiset.lcm_dvd,
+  simp only [multiset.mem_map, and_imp, exists_imp_distrib],
+  exact ⟨λ k b hb, k _ _ hb rfl, λ k a' b hb h, h ▸ k _ hb⟩,
+end
+
+lemma lcm_dvd {a : α} : (∀b ∈ s, f b ∣ a) → s.lcm f ∣ a :=
+lcm_dvd_iff.2
+
+lemma dvd_lcm {b : β} (hb : b ∈ s) : f b ∣ s.lcm f :=
+lcm_dvd_iff.1 (dvd_refl _) _ hb
+
+@[simp] lemma lcm_insert [decidable_eq β] {b : β} :
+  (insert b s : finset β).lcm f = gcd_monoid.lcm (f b) (s.lcm f) :=
+begin
+  by_cases h : b ∈ s,
+  { rw [insert_eq_of_mem h,
+    (lcm_eq_right_iff (f b) (s.lcm f) (multiset.normalize_lcm (s.1.map f))).2 (dvd_lcm h)] },
+  apply fold_insert h,
+end
+
+@[simp] lemma lcm_singleton {b : β} : ({b} : finset β).lcm f = normalize (f b) :=
+lcm_singleton
```

Comment:
This definition is a bit confusing because it looks self-referential. Even though `multiset` is open, I prefer:
```suggestion
multiset.lcm_singleton
```